### PR TITLE
planner: make tidb_enable_plan_cache_for_subquery can take effective immediately

### PR DESCRIPTION
**Pull request by fzzf678**
### What problem does this PR solve?

Issue Number: ref #40219

Problem Summary:

### What is changed and how it works?
Make the switch effective after `prepare` and `execute`
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/plan_cache_utils.go [hunk 1]
```diff
@@ -470,44 +470,48 @@ func (checker *limitExtractor) Enter(in ast.Node) (out ast.Node, skipChildren bo
...
 
 // Leave implements Visitor interface.
 func (checker *limitExtractor) Leave(in ast.Node) (out ast.Node, ok bool) {
-	return in, checker.cacheable
+	return in, checker.cacheable || !checker.hasSubQuery
 }
 
 // ExtractLimitFromAst extract limit offset and count from ast for plan cache key encode
-func ExtractLimitFromAst(node ast.Node, sctx sessionctx.Context) ([]uint64, error) {
+func extractLimitAndEstimateSubQueryFromAst(node ast.Node, sctx sessionctx.Context) ([]uint64, bool, error) {
```
Comment (qw4990) [inline planner/core/plan_cache_utils.go:485]: How about renaming this function to `extractMatchOptsFromAST(...) (*PlanCacheMatchOpts, bool)`?

File: planner/core/plan_cache_utils.go [hunk 2]
```diff
@@ -470,44 +470,48 @@ func (checker *limitExtractor) Enter(in ast.Node) (out ast.Node, skipChildren bo
...
 	if node == nil {
-		return nil, nil
+		return nil, false, nil
 	}
 	checker := limitExtractor{
```
Comment (qw4990) [inline planner/core/plan_cache_utils.go:489]: How about renaming this to `matchOptsExtractor`?

File: planner/core/plan_cache_utils.go (hunk 1 above)
Comment (fzzf678) [reply planner/core/plan_cache_utils.go]: Updated, PTAL again

**Review by Reminiscent — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**